samza-yarn/src/main/scala/org/apache/samza/job/yarn/SamzaYarnAppMasterLifecycle.scala

```diff
@@ -68,8 +71,17 @@ class SamzaYarnAppMasterLifecycle(containerMem: Int, containerCpu: Int, samzaApp
     //allowing the RM to restart it (potentially on a different host)
     if(samzaAppStatus != SamzaAppStatus.UNDEFINED) {
       info("Unregistering AM from the RM.")
-      amClient.unregisterApplicationMaster(yarnStatus, shutdownMessage, null)
-      info("Unregister complete.")
+      try {
+        amClient.unregisterApplicationMaster(yarnStatus, shutdownMessage, null)
+        info("Unregister complete.")
+      } catch {
+        case ex: InvalidApplicationMasterRequestException =>
+          // Once the AM dies, the corresponding app attempt ID is removed from the RM cache so that the RM can spin up a new AM and its containers.
```

Comment:
AM isn't dead yet right? Shouldn't this be `once the NM dies` Which brings up another interesting question. Have you validated the behavior when `yarn.nodemanager.recovery.enabled` is set to `true` vs `false`?

Reply:
Fixed the log. Also validated with `yarn.nodemanager.recovery.enabled` set to `true`. It works as expected.